community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java

```diff
@@ -120,147 +141,93 @@ public PropertyItem get()
+                    for ( int i = 0; remainingBlocksToRead-- > 0; i++ )
+                    {
+                        valuesAfterHeader[i] = cursor.getLong();
+                    }
                 }
-                break;
+                while ( cursor.shouldRetry() );
             }
-
-            case FLOAT:
+            catch ( IOException e )
             {
-                buffer.putInt( (int) header );
-                break;
+                throw new UnderlyingStorageException( e );
             }
+        }
+    }
 
-            case DOUBLE:
-            {
-                buffer.putLong( cursor.getLong() );
-                remainingBlocksToRead--;
-                break;
-            }
+    private void readFromStore( AbstractDynamicStore store ) throws IOException
+    {
+        int storeOffset = cursor.getOffset();
+        cursor.close();
+        cursor = null;
 
-            case STRING:
-            {
-                int storeOffset = cursor.getOffset();
-                cursor.close();
+        if ( recordCursor == null )
+        {
+            recordCursor = store.newDynamicRecordCursor();
+        }
 
-                if ( stringRecordCursor == null )
-                {
-                    stringRecordCursor = stringStore.newDynamicRecordCursor();
-                }
+        buffer.clear();
+        long startBlockId = (header & 0xFFFFFFFFF0000000L) >>> 28;
+        try ( GenericCursor<DynamicRecord> records = store.getRecordsCursor( startBlockId, true, recordCursor ) )
+        {
+            while ( records.next() )
+            {
+                DynamicRecord dynamicRecord = records.get();
 
-                try ( GenericCursor<DynamicRecord> stringRecords = stringStore.getRecordsCursor( header, true,
-                        stringRecordCursor ) )
+                while ( true )
                 {
-                    while ( stringRecords.next() )
+                    try
                     {
-                        DynamicRecord dynamicRecord = stringRecords.get();
-
                         buffer.put( dynamicRecord.getData(), 0, dynamicRecord.getData().length );
+                        break;
                     }
-                }
-
-                cursor = propertyStore.newReadCursor( currentRecordId );
-                cursor.setOffset( storeOffset );
-                break;
-            }
-
-            case ARRAY:
-            {
-                int storeOffset = cursor.getOffset();
-                cursor.close();
-
-                if ( arrayRecordCursor == null )
-                {
-                    arrayRecordCursor = arrayStore.newDynamicRecordCursor();
-                }
-
-                try ( GenericCursor<DynamicRecord> arrayRecords = arrayStore.getRecordsCursor( header, true,
-                        arrayRecordCursor ) )
-                {
-                    while ( arrayRecords.next() )
+                    catch ( BufferOverflowException e )
```

Comment:
Exceptions for flow control :(
Maybe we can use `ByteBuffer#remaining` or even `ByteArrayOutputStream` that resized automatically?